### PR TITLE
refactor(facade): use model config service in provisioner facade

### PR DESCRIPTION
**Pull request by barrettj12**
This PR refactors the provisioner facade and the common ToolsGetter to use model services. The introduced services are model info/config/agent as appropriate.

Additional changes:
- Remove the unused configGetter from common.ToolsFinder.
- Rename `stdcontext` to `context` in some places.
- Be safer when casting the value in `Config.SSLHostnameVerification` to avoid panics.
- Saving machine data to dqlite should be idempotent.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- ~[ ] Comments saying why design decisions were made~
- [ ] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Affected facades and methods are:
- Provisioner: GetContainerProfileInfo, ModelUUID, ProvisioningInfo
- Upgrader: Tools

To test the provisioner, bootstrap Juju on AWS:
```
$ juju bootstrap aws aws
$ juju add-model m
$ juju add-machine
created machine 0
$ juju deploy ubuntu --to lxd:0
```
Check that ubuntu comes up correctly and doesn't crash the container or the host machine.

Set the resource tags in model config:
```
juju model-config resource-tags="foo=bar"
```

Create a storage pool:
```
juju create-storage-pool ebs ebs size=1G
```

Provision a new machine with an attached volume:
```
juju add-machine --disks="ebs,100M,1"
```
and check on the AWS console that the provisioned machine and attached volume both have the specified resource tags.

**TODO:** test upgrader worker

## Links

**Jira card:** [JUJU-6056](https://warthogs.atlassian.net/browse/JUJU-6056)

[JUJU-6056]: https://warthogs.atlassian.net/browse/JUJU-6056?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 1]
```diff
@@ -378,7 +387,7 @@ func (api *ProvisionerAPI) ContainerManagerConfig(ctx stdcontext.Context, args p
 	cfg := make(map[string]string)
 	cfg[container.ConfigModelUUID] = api.st.ModelUUID()
 
-	mConfig, err := api.m.ModelConfig(ctx)
+	mConfig, err := api.modelConfigService.ModelConfig(ctx)
```
Comment (dimaqq) [inline apiserver/facades/agent/provisioner/provisioner.go:390]: If the config is pulled from the service now, should it disappear from `api.m`?

Comment (nvinuesa) [reply apiserver/facades/agent/provisioner/provisioner.go:390]: The `m` struct is the model that lives in the legacy state, so no point in removing stuff from it (we are removing it soon-ish). Furthermore, the full model hasn't still been migrated to dqlite so we can't fully remove this field from the `api`.

**Comment by nvinuesa** [comment]
This facade is used in the `containerbroker` and `provisioner` (container init) workers. I'd QA it by adding some container machines and deploying into it, something like:
```
juju add-machine lxd
juju deploy --to lxd:0
```
Also, I'm a bit surprised that tests pass. I would expect something to fail with nil point reference at this point (since you haven't added anything to any test), otherwise it means that we aren't testing that path.

File: apiserver/facades/agent/provisioner/provisioner.go (hunk 1 above)
Comment (tlm) [reply apiserver/facades/agent/provisioner/provisioner.go:390]: Normally in these PR's we would do this. But this facade isn't using an interface for the model state so we can't remove it yet.

**Comment by tlm** [comment]
There is a few other things to touch up in this file as we go along.

`provisioner.go`
```go
func (api *ProvisionerAPI) ModelUUID(ctx stdcontext.Context) params.StringResult {
	return params.StringResult{Result: api.st.ModelUUID()}
}
```
Can we please change this over to use the model info service (https://github.com/juju/juju/blob/main/domain/model/service/modelservice.go).

**Review by tlm — COMMENTED** [review]
Thank you for the work. When you are ready for a final review can you please update with testing instructions. Also please bring the card into the current pulse.

**Review by manadart — COMMENTED** [review]
This is OK, but it's work that will need double handling.

We want to change the model manager workers (of which this is one) so that they use the domain services directly instead of going via the API.

The work has already been done to supply the service factory to the manifolds configuration, so if we do any more of these we might as well just stop using the API methods and go via the service.

**Comment by barrettj12** [comment]
I see a potential complication that this facade is used in multiple places.

For example, the `ContainerManagerConfig` method is used by the `provisioner` worker and the `containerbroker` worker. Ideally we'd want to avoid copy-pasting this method - so would we make one worker depend on the other, or put this code in a "common" package that both could import, or something else?

The `ContainerConfig` method is used by the LXD container broker. Is it (or will it be) possible to supply the model/controller config services to this broker?

And I think `ProvisioningInfo` still depends too much on state to be moved into a worker

**Comment by manadart** [comment]
The strategy is to push logic from the facades into the service layer so that to the greatest extent possible, facades only deal with:
- auth
- encoding/decoding wire types
- dispatch to the services

With this strategy there is no issue with copy/pasting or duplication, because both sites ultimately recruit the same service method with the API facade being... a facade.

You won't be able to use services from the container broker directly, as it resides on leaf agents.
You don't need to consider `ProvisioningInfo` at this stage, only direct calls to config watching/reading.

**Review by manadart — COMMENTED** [review]
Something is broken between this HEAD.

If you:
- `juju add-machine`
- `juju deploy ubuntu --to lxd:0`

It works on HEAD, but will hard freeze the host machine with this patch, status reporting "agent lost".

**Comment by barrettj12** [comment]
In debug-log, I see the lines:
```
machine-0: 13:09:17 ERROR juju.worker.dependency "broker-tracker" manifold worker returned unexpected error: no container types determined
machine-0: 13:09:17 ERROR juju.worker.dependency "lxd-container-provisioner" manifold worker returned unexpected error: container types not yet available
```

Both of these error messages can be traced back to the `SupportedContainers` method, it seems I've broken that somehow.

**Comment by manadart** [comment]
I don't think that explains the loss of the agent. Remember, this works on `HEAD`.

**Comment by hpidcock** [comment]
> I don't think that explains the loss of the agent. Remember, this works on `HEAD`.

So I have spent a lot of time diagnosing this.
TLDR: The root of the problem is that we are not auto-configuring "container-networking-method" in model config.

The long story: The EC2 instance loses network connectivity. Diagnosing this was a right pain, but I had to add the machine first, connect to it over ssh, set a password for the `ubuntu` user, then connect via the EC2 serial console and login with that password.

Why didn't it have network connectivity after the lxd container began provisioning? 
The reason the EC2 instance lost connectivity, is because a bridge interface was being created with a random mac address (replacing the original interface), this meant that the VPC DHCP was unable to respond to an unknown mac address and allocate addresses to it. 
A solution here was to patch `internal/network/netplan/netplan.go` to use the mac address of the device being matched by the original interface.

Juju was never meant to be creating a bridge interface on EC2 (because AWS doesn't, at least out of the box, support DHCP for this). This led me to `container-networking-method` model config value being the root cause of this failure, setting to `local` after creating the model solves the issue.

As for bridge networking, it probably makes sense to do something to the effect of this patch, as other environments might have similar issues.

internal/network/netplan/netplan.go patch:
```
// BridgeEthernetById takes a deviceId and creates a bridge with this device
// using this devices config
func (np *Netplan) BridgeEthernetById(deviceId string, bridgeName string) (err error) {
	ethernet, ok := np.Network.Ethernets[deviceId]
	if !ok {
		return errors.NotFoundf("ethernet device with id %q for bridge %q", deviceId, bridgeName)
	}
	shouldCreate, err := np.shouldCreateBridge(deviceId, bridgeName)
	if !shouldCreate {
                // err may be nil, but we shouldn't continue creating
                return errors.Trace(err)
        }
-       np.createBridgeFromInterface(bridgeName, deviceId, &ethernet.Interface)
+       // If the ethernet device was being matched on mac but did not specify a mac address for
+       // the interface, we need to copy the matched mac address to the bridge interface macaddress
+       // field, otherwise on AWS the bridge interface will not get addresses from dhcp.
+       ethInf := ethernet.Interface
+       macaddressSet := false
+       if mac, ok := ethernet.Match["macaddress"]; ok && ethInf.MACAddress == "" {
+               macaddressSet = true
+               ethInf.MACAddress = mac
+       }
+       np.createBridgeFromInterface(bridgeName, deviceId, &ethInf)
+       if macaddressSet {
+               ethInf.MACAddress = ""
+       }
+       ethernet.Interface = ethInf
        np.Network.Ethernets[deviceId] = ethernet
        return nil
 }
```

Before the patch:
```
ubuntu@ip-172-31-54-202:/etc/netplan$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp39s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq master br-enp39s0 state UP group default qlen 1000
    link/ether 02:11:27:5e:dd:a7 brd ff:ff:ff:ff:ff:ff
3: br-enp39s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
    link/ether 52:a0:79:ef:19:39 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::50a0:79ff:feef:1939/64 scope link 
       valid_lft forever preferred_lft forever
```

After the patch:
```
ubuntu@ip-172-31-54-202:/etc/netplan$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp39s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq master br-enp39s0 state UP group default qlen 1000
    link/ether 02:11:27:5e:dd:a7 brd ff:ff:ff:ff:ff:ff
3: br-enp39s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
    link/ether 02:11:27:5e:dd:a7 brd ff:ff:ff:ff:ff:ff
    inet 172.31.54.202/20 metric 100 brd 172.31.63.255 scope global dynamic br-enp39s0
       valid_lft 3597sec preferred_lft 3597sec
    inet6 fe80::50a0:79ff:feef:1939/64 scope link 
       valid_lft forever preferred_lft forever
```

**Comment by hpidcock** [comment]
@manadart perhaps we force `container-networking-method` to `local` until we get a chance to reimplement the auto configuration?

**Comment by manadart** [comment]
> @manadart perhaps we force `container-networking-method` to `local` until we get a chance to reimplement the auto configuration?

I will look at this during the current pulse.

**Comment by tlm** [comment]
@manadart we have already re-implemented auto config for container networking. Checkout #17574

File: domain/modelagent/service/service.go [hunk 2]
```diff
@@ -41,6 +43,19 @@ func (s *Service) GetModelAgentVersion(ctx context.Context, modelID model.UUID)
 	return s.st.GetModelAgentVersion(ctx, modelID)
 }
 
+// SetModelAgentVersion sets the agent version for the specified model.
+// The following errors can be returned:
+//   - [errors.NotValid] if the model ID is not valid.
+//
+// NB: this method was only implemented for testing purposes. You should be
+// very careful and think twice before using this in production
+func (s *Service) SetModelAgentVersion(ctx context.Context, modelID model.UUID, ver version.Number) error {
```
Comment (manadart) [inline domain/modelagent/service/service.go:52]: Don't do this. It's only used in one test, so get the transaction running from the suite, and execute some SQL to do it.

**Review by manadart — COMMENTED** [review]
Some changes to make.

File: domain/modelagent/service/service.go (hunk 2 above)
Comment (barrettj12) [reply domain/modelagent/service/service.go:52]: We need it for the jujud-controller agent tests too. That's really the only reason we implemented it.
https://github.com/barrettj12/juju/blob/1f13d1259d75c5a523a24b8457ef6bdae49f3998/cmd/jujud-controller/agent/machine_test.go#L268

File: apiserver/common/service.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/core/credential"
+	"github.com/juju/juju/core/crossmodel"
+	"github.com/juju/juju/core/machine"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/environs/config"
+)
+
+// ModelAgentService provides access to the Juju agent version for the model.
+type ModelAgentService interface {
```
[the remark this answers is not in the document]
Comment (barrettj12) [reply apiserver/common/service.go:21]: right, but the interface definitions are package-scoped. If two of these "unrelated common modules" both need to access a model config service, they won't be able to define separate interfaces. Hence IMO it makes sense to have the service interfaces centrally defined.

[the remark this answers is not in the document]
Comment (manadart) [reply apiserver/common/service.go:21]: If they are unrelated, they should define their own interfaces. Other wise said interface has > 1 reason to change I.e SRP.

Revert it please.

File: domain/modelagent/service/service.go (hunk 2 above)
Comment (manadart) [reply domain/modelagent/service/service.go:52]: Put in test code as I've described.

**Review by manadart — CHANGES_REQUESTED** [review]
2 things to turn around.

File: apiserver/common/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (barrettj12) [reply apiserver/common/service.go]: OK reverted

File: domain/modelagent/service/service.go (hunk 2 above)
Comment (barrettj12) [reply domain/modelagent/service/service.go]: done

**Review by manadart — APPROVED** [review]
All looking good. Thanks for the turn-around.
